### PR TITLE
 Added bootstrap feature

### PR DESCRIPTION
**Pull request by rawmind0**
This PR add bootstrap feature.
Bootstrap is doing following tasks:
- Update default admin password, provided by `password` or generating a random one.
- Set `server-url` setting, based on `api_url`.
- Set `telemetry-opt` setting.
- Create a token for admin user with concrete TTL.
- Set `token_update` setting to regenerate admin token.
- Added bootstrap resource tests.

Admin password could be updated setting `password` field and applying this resource again. Login to rancher is done using `token` first and if fails, using admin `current_password`. If admin password has been changed from other methods and terraform token is expired, `current_password` field could be especified to recover terraform configuration and reset admin password and token.

### CODE REVIEW

File: rancher2/util.go [hunk 1]
```diff
@@ -21,7 +21,7 @@ import (
 const (
 	clusterProjectIDSeparator = ":"
 	passDigits                = "abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ0123456789~!@#$%^&*()_+`-={}|[]\\:\"<>?,./"
-	passDefaultLen            = 16
+	passDefaultLen            = 20
```
Comment (cloudnautique) [inline rancher2/util.go:25]: Why the increase?

Comment (rawmind0) [reply rancher2/util.go:25]: Just to set same length than [resetpassword](https://github.com/rancher/rancher/blob/master/app/resetpassword.go)

File: website/docs/r/bootstrap.html.markdown [hunk 2]
```diff
@@ -0,0 +1,52 @@
...
+# Provider config
+provider "rancher2" {
+  api_url   = "https://rancher.my-domain.com"
+  bootstrap = true
+}
+
+# Create a new rancher2 Bootstrap
+resource "rancher2_bootstrap" "admin" {
+  password = "blahblah"
+  telemtry = true
```
Comment (drpebcak) [inline website/docs/r/bootstrap.html.markdown]: Should say 'telemetry'

Comment (rawmind0) [reply website/docs/r/bootstrap.html.markdown]: fixed

**Comment by drpebcak** [comment]
I did notice that you can only set the admin password once, and then any changes to it afterwards fail silently. Is it possible for us to use the api key once we have it to allow password resets?

**Comment by rawmind0** [comment]
This resource is just to bootstrap rancher, indeed to be run once. One of the tasks is change default password, then the provider is using api keys to manage the rest of the resources. btm, it doesn't support change users password.

**Comment by drpebcak** [comment]
Is there some limitation why you can't change the admin user's password more than once? The biggest issue here is that terraform sees this as a valid operation, so if you run `plan`, it says it will modify it.

**Comment by rawmind0** [comment]
User password can be updated more than once, but the current values couldn't be got from API, then they can't be compared and checked from rancher, just from tfstate values.

Otherwise, bootstrap is a "special" resource. The bootstrap mode is scoped to be run just once (first login is using default admin user/password) generating its own tfstate, it bootstraps rancher system (change default pass, set url, set telemetry option and generate a token) and output values to be used to manage other provider resources. Current implementation is generating an error if `rancher2_bootstrap` resource exists and bootstrap mode is not set.

```
	if !meta.(*Config).Bootstrap {
		return fmt.Errorf("[ERROR] Resource rancher2_bootstrap just available on bootsrap mode")
	}
```

May be we can change the scope and make that if bootstrap flag is set and `rancher2_bootstrap` resource is updated (compared with tfstate), just update changes and not made full bootstrap process.

**Comment by cloudnautique** [comment]
If we used the kubectl method you could compare them from the crd, possibly?

**Comment by cloudnautique** [comment]
Also, the bootstrap resource is supposed to be special and limited in scope, but it is meant to be a functioning resource over the life of the Rancher server.

The API tokens for instance can be passed to a down stream TF workflow for managing the auth providers and other admin/owner level constructs. So, I think we need an answer for the password/api key changes.

**Comment by drpebcak** [comment]
Just to clarify, if we have a terraform module that:

1. Provisions a k8s cluster
2. Installs rancher
3. Uses this bootstrap resource to set admin password, server-url, and telemetry options

We will want to run this multiple times - if there is a k8s upgrade, rancher upgrade, etc. It is ok if I can't use any of the other rancher2 resources in this module, but I should expect to be able to make changes here and see them applied.

**Comment by rawmind0** [comment]
Upgrade bootstrap resource to be able to be run multiple times and update admin password. Token is regenerated if you add param `token_update = true`, removing previous one. 

Also, added `current_password` field just to be able to recover if admin password has been changed from other method and `token` is expired. The login to rancher is done using `token` first and if fails, using admin `current_password`.

Please take a look...

**Comment by rawmind0** [comment]
Updated PR adding tests for bootstrap resource and adding `token_update` (bool) param to regenerate admin token.

**Comment by cloudnautique** [comment]
To clarify, the token_update is only on admin password change?

**Comment by rawmind0** [comment]
Not really, if `token_update = true` token is regenerated, not needed admin password change.

File: rancher2/resource_rancher2_bootstrap.go [hunk 3]
```diff
@@ -0,0 +1,266 @@
...
+		"telemetry": &schema.Schema{
+			Type:     schema.TypeBool,
+			Optional: true,
+			Default:  false,
+		},
+		"url": &schema.Schema{
+			Type:     schema.TypeString,
+			Computed: true,
+		},
+		"user": &schema.Schema{
```
Comment (nazarewk) [inline rancher2/resource_rancher2_bootstrap.go:75]: `user` is defined, but not `.Set()` anywhere

**Comment by rawmind0** [comment]
Updated PR to check if token is expired. On refresh function, if token is expired, it set `token_update = true` to force token update on next `terraform apply` 

Also added cleanup function to remove login tokens, that are generated if connected with user/pass.

File: rancher2/resource_rancher2_bootstrap.go (hunk 3 above)
Comment (rawmind0) [reply rancher2/resource_rancher2_bootstrap.go:75]: thanks @nazarewk, you are right. Fixed.

**Comment by drpebcak** [comment]
Looks like this works as expected 👍  - the only thing that is a bit weird, is that if I delete the token through the UI, terraform will recreate it for me, except it no longer has the `Terraform bootstrap admin token` description. It's just blank.

**Comment by cloudnautique** [comment]
@rawmind0 can we get the description added to the token, and then we can merge this.

**Comment by rawmind0** [comment]
@drpebcak , it's not the same token. This token is generated when the resource is logged in using user/pass, called `temp_token` in the resource.  This token has a ttl of 60s and should be removed once the admin token is generated.

Anyway, i've added a description for this token `Terraform bootstrap admin session` to differentiate from the other.

**Comment by drpebcak** [comment]
Ah, I see. In that case, it seems like terraform is unaware that the token it generated is now gone. It keeps the bogus token as an output. Is there any way we can resolve that so that the user is not left without a token?

**Comment by rawmind0** [comment]
May be you are not testing latest version, the resource already do that. If the generated token is removed from ui or is expired, resource will generate a new token on next `terraform apply` execution.

**Pull request merged**